### PR TITLE
Add custom two,three and four terminal devics

### PR DESCRIPTION
**Pull request by RehabSayed-G**
In this PR, we will make the following changes:
    * Add two terminal custom device and test it with heater doped device
    * Add N terminal custom device and test it with mmi1x2,mmi2x2 and mmi_90_dgree_hybrid devices 
    * Update devices previously implemented with Resistor class to use the general n_terminal class

### CODE REVIEW

**Comment by RehabSayed-G** [comment]
@FaragElsayed2 , 
Ready For Review for the following devices : 
* straight_heater_metal
* straight_heater_metal_undercut
* mzi_mmi
* mmi1x2
* mmi2x2
* mmi_90_degree_hybrid

File: klayout/lvs/rule_decks/general_derivations.lvs [hunk 1]
```diff
@@ -1,3 +1,5 @@
 #==================================
 # ------ GENERAL DERIVATIONS ------
 #==================================
+
+wg_con = wg.not(heater_mk).not(heater_length_mk).not(mmi_mk).not(mmi_length_mk).not(mzi_len_mk)
```
Comment (FaragElsayed2) [inline klayout/lvs/rule_decks/general_derivations.lvs:5]: Add a new exclusion layer then use it, for optimization purpose

```
wg_con_exclude = heater_mk.join(heater_length_mk).join(mmi_mk).join(mmi_length_mk).join(mzi_len_mk)

wg_con = wg.not(wg_con_exclude)
```

File: klayout/lvs/rule_decks/heater_extraction.lvs [hunk 2]
```diff
@@ -10,18 +10,42 @@ logger.info('Starting HEATER EXTRACTION')
...
+                                                                            })
+
+# straight_heater_metal_undercut
+logger.info('Extracting straight_heater_metal_undercut device')
+extract_devices(GeneralNTerminalExtractor::new('straight_heater_metal_undercut',2), {"seed" => straight_heater_metal_undercut_res, 
+                                                                            "contact"=> straight_heater_metal_undercut_con,
+                                                                            "measure_mk" => straight_heater_metal_undercut_res
+                                                                            })
+
+# # ====================
```
Comment (FaragElsayed2) [inline klayout/lvs/rule_decks/heater_extraction.lvs:25]: To be modified as we agreed for other devices [Using custom class - N term]

File: klayout/lvs/rule_decks/heater_extraction.lvs [hunk 3]
```diff
@@ -10,18 +10,42 @@ logger.info('Starting HEATER EXTRACTION')
...
+# # --- HEATER DOPED ---
```
Comment (FaragElsayed2) [inline klayout/lvs/rule_decks/heater_extraction.lvs:26]: To be modified as we agreed for other devices [Using custom class - N term]

File: klayout/lvs/testing/testcases/unit/heater_devices/netlist/straight_heater_doped/straight_heater_doped_rib.spice
Comment (FaragElsayed2) [inline klayout/lvs/testing/testcases/unit/heater_devices/netlist/straight_heater_doped/straight_heater_doped_rib.spice:1]: Please add Licence for all netlists.

File: klayout/lvs/rule_decks/custom_classes.lvs [hunk 4]
```diff
@@ -9,7 +9,9 @@ class SubcircuitModelsReader < RBA::NetlistSpiceReaderDelegate
   rescue StandardError
     case element
     when 'Z'
-      super("#{sup} lx=0 ly=0 dl=0", element)      
+      super("#{sup} lx=0 ly=0 dl=0", element)
```
Comment (FaragElsayed2) [inline klayout/lvs/rule_decks/custom_classes.lvs:12]: We don't have any `Z` or `R` devices, could you check this?

File: klayout/lvs/rule_decks/custom_classes.lvs [hunk 5]
```diff
@@ -9,7 +9,9 @@ class SubcircuitModelsReader < RBA::NetlistSpiceReaderDelegate
   rescue StandardError
     case element
     when 'Z'
-      super("#{sup} lx=0 ly=0 dl=0", element)      
+      super("#{sup} lx=0 ly=0 dl=0", element)    
+    when 'R'
+      super("#{sup} R=0", element)
```
Comment (FaragElsayed2) [inline klayout/lvs/rule_decks/custom_classes.lvs:14]: Same as above

File: klayout/lvs/rule_decks/custom_classes.lvs [hunk 6]
```diff
@@ -19,29 +21,106 @@ class SubcircuitModelsReader < RBA::NetlistSpiceReaderDelegate
...
+    
+        # create a device
+        device = circuit.create_device(cls, name)
+    
+        # and configure the device
+        %w[1 2 3].each_with_index do |t, index|
+          device.connect_terminal("TERMINAL_#{t}", nets[index])
+        end
+
+        if model.include?('MMI1X2')
```
Comment (FaragElsayed2) [inline klayout/lvs/rule_decks/custom_classes.lvs:69]: Why is this specific for `mmi1x2`? it could be used for all `mmi` types.

Comment (FaragElsayed2) [inline klayout/lvs/rule_decks/custom_classes.lvs:69]: You could combine the following checks outside `nets.size` check.

```
if model.include?('MMI1X2')
if model.include?('MMI2X2')
if model.include?('MMI_90DEGREE')
```

File: klayout/lvs/rule_decks/custom_classes.lvs [hunk 7]
```diff
@@ -19,29 +21,106 @@ class SubcircuitModelsReader < RBA::NetlistSpiceReaderDelegate
...
-  
-      # create a device
-      device = circuit.create_device(cls, name)
-  
-      # and configure the device
-      %w[A B].each_with_index do |t, index|
-        device.connect_terminal(t, nets[index])
+
+      if nets.size == 2
+        unless cls
```
Comment (FaragElsayed2) [inline klayout/lvs/rule_decks/custom_classes.lvs:29]: Device creation could be done only one time outside the `nets.size` check for all conditions.

File: klayout/lvs/rule_decks/custom_classes.lvs [hunk 8]
```diff
@@ -217,7 +641,7 @@ end
 class NResistor < RBA::DeviceClassResistor
```
Comment (FaragElsayed2) [inline klayout/lvs/rule_decks/custom_classes.lvs:641]: I believe we couldn't use this anymore.

File: klayout/lvs/rule_decks/custom_classes.lvs [hunk 9]
```diff
@@ -202,6 +286,346 @@ class MZIExtractor < RBA::GenericDeviceExtractor
   end
 end
 
+class DeviceClassNTerminal < RBA::DeviceClassResistor
```
Comment (FaragElsayed2) [inline klayout/lvs/rule_decks/custom_classes.lvs:289]: Could you please split all custom devices in new lvs file, then include it in this file or in the main one?

File: klayout/lvs/rule_decks/custom_classes.lvs
Comment (FaragElsayed2) [inline klayout/lvs/rule_decks/custom_classes.lvs:1]: Please make sure to add comments in this file to explain your steps. Also please remove any code comments.

**Review by FaragElsayed2 — CHANGES_REQUESTED** [review]
Please check required changes @RehabSayed-G

**Comment by RehabSayed-G** [comment]
@FaragElsayed2 , Ready for review

**Review by FaragElsayed2 — APPROVED** [review]

**Pull request merged**